### DISCUSSION TITLE
Coordinate System/ Units restrictions

### DISCUSSION THREAD

**Discussion by JamesZDonline** [PUTvision/qgis-plugin-deepness#109] [Q&A]
First let me start by saying thank you for your great work on this tool. I'm extremely excited about it and can see how it can fit well into my existing workflows.

I was wondering about restrictions in the coordinate systems of the imagery . It appears the plugin assumes layer units are in meters. I tried running a model from the model zoo on some WGS84 (EPSG: 4326) imagery and it gave me the following error:

`File "C:\Users/zimmejr1/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\deepness\processing\processing_utils.py", line 31, in convert_meters_to_rlayer_units
              raise Exception("Unsupported layer units")
             Exception: Unsupported layer units`

Is there any way around this other than re-projecting imagery? It works as expected when I use a pseudomercator projection, but I have a lot of imagery and require EPSG 4326 for other parts of my workflow so I can't just re-project it all.

**Comment by przemyslaw-aszkowski** [comment]
Hi @JamesZDonline,

Thanks for your feedback. It is possible that Deepness doesn't support some units at the moment. However, I expect it should be possible to add support for that. If you are interested, please provide example images and instructions on how to reproduce :)

Cheers
Przemek

**Reply by JamesZDonline** [reply]
Hello @przemyslaw-aszkowski 

Thank you for your response. That sounds great! I am providing a google link with a 1.3GB example image. If you would like me to provide the data in another way, or if there's anything else I can do please let me know. 
[Image](https://drive.google.com/file/d/1DztkSz0Yfoy84UVsx3AG2BFNW3ky5A4V/view?usp=drive_link)

For instructions to reproduce, I did the following
1. Opened deepness plugin
2. Selected my image
3. Selected Processed area mask as "Visible part"
4. Used any landcover or object detection onnx model from the Deepness model zoo
5. Defined an output folder
6. Clicked run

All the best,
James

**Comment by przemyslaw-aszkowski** [comment]
Hi James,
Thanks for providing the details.

At the moment Deepness internal processing needs to convert between meters and raster layer units in order to know the resolution. Your orthophoto has degree units, and it causes the issue.

After a quick research, I'm not sure if it is simple to convert between meters and degrees, as they are based on different projection systems. If you would like to investigate, here is the relevant code in our plugin:
```
def convert_meters_to_rlayer_units(rlayer: QgsRasterLayer, distance_m) -> float:
    """ How many map units are there in one meter """
    # TODO - potentially implement conversions from other units
    if rlayer.crs().mapUnits() != QgsUnitTypes.DistanceUnit.DistanceMeters:
        # TODO - add support for more unit types
        raise Exception("Unsupported layer units")
    assert distance_m != 0
    return distance_m
```
I suspect it would be possible to implement it (and definitively good to have in the Deepness!), though we may not have time to do it now, so any help welcome!

But I found a quite **simple workaround** - you just need to reproject your files:
1) right click on your raster layer
2) Click Export->Save As
3) type some file name and select CRS with meter units, e.g. EPSG:3857
4) click OK and wait a few seconds.
5) Now the new file contains the same data, though reprojected, and can be processed with Deepness.

Hope it helps!
Przemek

**Reply by JamesZDonline** [reply]
Hello,

Thank you for looking into this!

If I am understanding the code correctly, it looks like the only purpose of convert_meters_to_rlayer_units is to then calculate the units per pixel? Is this correct? 

If so, I think there may be a simple solution. QGSRasterLayer has methods to retrieve the units per pixel directly:

QGSRasterLayer.rasterUnitsPerPixelX() 
QGSRasterLayer.rasterUnitsPerPixelY()

This will work for any coordinate system and should bypass the need for converting between meters and degrees.

If this seems plausible to you I can test it and try creating a pull request. I've never contributed to a Github project before so I don't know the proper etiquette and procedure but am happy to contribute if you're willing to be patient with me.


We have about 15TB of imagery, so re-projecting it all is not practical. I'm doing the majority of the model training and inference in RasterVision, but am excited about the possibility to use Deepness to rapidly spot-check small areas with new model versions and as a tool for rapid deployment for the training data collection team.

Thanks again and let me know if this seems worth pursuing!

All the best,
James

**Reply by przemyslaw-aszkowski** [reply]
Hi James,
After looking at the code I believe your understanding is correct, and the functions you mentioned also seem reasonable. 
One thing I'm not sure about is how to handle the situation when the units per pixel in the x and y axes are different (some processing code may need refactorization to handle both axes individually).

Of course, I will appreciate your contribution :) The easiest way to contribute is to fork the repository, and then create a pull request to the parent repository. I recommend having a look at the development instructions in the main README. Let me know if you need any support with development.

Cheers,
Przemek

**Comment by przemyslaw-aszkowski** [comment]
Hi
Should be resolved now with [version 0.6.5](https://github.com/PUTvision/qgis-plugin-deepness/releases/tag/0.6.5)
